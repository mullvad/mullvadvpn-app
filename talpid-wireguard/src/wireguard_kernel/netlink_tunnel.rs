use std::pin::Pin;

use futures::Future;

use super::{
    super::stats::{Stats, StatsMap},
    wg_message::DeviceNla,
    Config, Error, Handle, Tunnel, TunnelError, MULLVAD_INTERFACE_NAME,
};

pub struct NetlinkTunnel {
    interface_index: u32,
    netlink_connections: Handle,
    tokio_handle: tokio::runtime::Handle,
}

impl NetlinkTunnel {
    pub fn new(tokio_handle: tokio::runtime::Handle, config: &Config) -> Result<Self, Error> {
        tokio_handle.clone().block_on(async {
            let mut netlink_connections = Handle::connect().await?;
            let interface_index = netlink_connections
                .create_device(MULLVAD_INTERFACE_NAME.to_string(), config.mtu as u32)
                .await?;

            let mut tunnel = Self {
                interface_index,
                netlink_connections,
                tokio_handle,
            };

            if let Err(err) = tunnel.setup(config).await {
                if let Err(teardown_err) = tunnel
                    .netlink_connections
                    .delete_device(interface_index)
                    .await
                {
                    log::error!(
                        "Failed to tear down WireGuard interface after failing to apply config: {}",
                        teardown_err
                    );
                }
                return Err(err);
            }

            Ok(tunnel)
        })
    }

    async fn setup(&mut self, config: &Config) -> Result<(), Error> {
        self.netlink_connections
            .wg_handle
            .set_config(self.interface_index, config)
            .await?;

        for tunnel_ip in config.tunnel.addresses.iter() {
            self.netlink_connections
                .set_ip_address(self.interface_index, *tunnel_ip)
                .await?;
        }

        Ok(())
    }
}

impl Tunnel for NetlinkTunnel {
    fn get_interface_name(&self) -> String {
        let mut wg = self.netlink_connections.wg_handle.clone();
        let result = self.tokio_handle.block_on(async move {
            let device = wg.get_by_index(self.interface_index).await?;
            for nla in device.nlas {
                if let DeviceNla::IfName(name) = nla {
                    return Ok(name);
                }
            }
            Err(Error::Truncated)
        });

        match result {
            Ok(name) => name.to_string_lossy().to_string(),
            Err(err) => {
                log::error!("Failed to deduce interface name at runtime, will attempt to use the default name. {}", err);
                MULLVAD_INTERFACE_NAME.to_string()
            }
        }
    }

    fn stop(self: Box<Self>) -> std::result::Result<(), TunnelError> {
        let Self {
            mut netlink_connections,
            interface_index,
            tokio_handle,
        } = *self;
        tokio_handle.block_on(async move {
            if let Err(err) = netlink_connections.delete_device(interface_index).await {
                log::error!("Failed to remove WireGuard device: {}", err);
                Err(TunnelError::FatalStartWireguardError)
            } else {
                Ok(())
            }
        })
    }

    fn get_tunnel_stats(&self) -> std::result::Result<StatsMap, TunnelError> {
        let mut wg = self.netlink_connections.wg_handle.clone();
        let interface_index = self.interface_index;
        self.tokio_handle.block_on(async move {
            let device = wg.get_by_index(interface_index).await.map_err(|err| {
                log::error!("Failed to fetch WireGuard device config: {}", err);
                TunnelError::GetConfigError
            })?;
            Ok(Stats::parse_device_message(&device))
        })
    }

    fn set_config(
        &self,
        config: Config,
    ) -> Pin<Box<dyn Future<Output = std::result::Result<(), TunnelError>> + Send + 'static>> {
        let mut wg = self.netlink_connections.wg_handle.clone();
        let interface_index = self.interface_index;
        Box::pin(async move {
            wg.set_config(interface_index, &config)
                .await
                .map_err(|err| {
                    log::error!("Failed to fetch WireGuard device config: {}", err);
                    TunnelError::SetConfigError
                })
        })
    }
}
