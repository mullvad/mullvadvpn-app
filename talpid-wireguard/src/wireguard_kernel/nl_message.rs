use super::parsers;
use byteorder::{ByteOrder, NativeEndian};
use netlink_packet_core::{
    NetlinkDeserializable, NetlinkHeader, NetlinkPayload, NetlinkSerializable,
};
use netlink_packet_utils::{
    nla::{Nla, NlaBuffer, NlasIterator},
    traits::{Emitable, Parseable},
    DecodeError,
};
use std::{ffi::CString, io::Write, mem};

#[derive(err_derive::Error, Debug)]
pub enum Error {
    #[error(display = "Family name too long")]
    FamilyNameTooLong,
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct NetlinkControlMessage {
    cmd: u8,
    version: u8,
    pub nlas: Vec<ControlNla>,
}

impl NetlinkControlMessage {
    pub fn get_netlink_family_id(name: CString) -> Result<Self, Error> {
        if name.as_bytes_with_nul().len() > (libc::GENL_NAMSIZ as usize) {
            return Err(Error::FamilyNameTooLong);
        }
        Ok(Self {
            nlas: vec![ControlNla::FamilyName(name)],
            cmd: libc::CTRL_CMD_GETFAMILY as u8,
            version: 1,
        })
    }
}

impl NetlinkSerializable for NetlinkControlMessage {
    fn message_type(&self) -> u16 {
        libc::GENL_ID_CTRL as u16
    }

    fn buffer_len(&self) -> usize {
        mem::size_of::<libc::genlmsghdr>() + self.nlas.as_slice().buffer_len()
    }

    fn serialize(&self, mut buffer: &mut [u8]) {
        let _ = buffer.write(&[self.cmd, self.version, 0u8, 0u8]).unwrap();
        self.nlas.as_slice().emit(buffer);
    }
}

impl From<NetlinkControlMessage> for NetlinkPayload<NetlinkControlMessage> {
    fn from(msg: NetlinkControlMessage) -> Self {
        NetlinkPayload::InnerMessage(msg)
    }
}

impl NetlinkDeserializable for NetlinkControlMessage {
    type Error = DecodeError;
    fn deserialize(
        _header: &NetlinkHeader,
        payload: &[u8],
    ) -> Result<NetlinkControlMessage, Self::Error> {
        // skip the genlmsghdr
        let (cmd, version) = parsers::parse_genlmsghdr(payload)?;
        let nla_buffer = &payload[mem::size_of::<libc::genlmsghdr>()..];
        let nlas = NlasIterator::new(nla_buffer)
            .map(|buffer| ControlNla::parse(&buffer?))
            .collect::<Result<Vec<_>, DecodeError>>()?;

        Ok(NetlinkControlMessage { nlas, cmd, version })
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum ControlNla {
    FamilyName(CString),
    FamilyId(u16),
    Unknown(u16, Vec<u8>),
}

impl Nla for ControlNla {
    fn value_len(&self) -> usize {
        use ControlNla::*;
        match self {
            FamilyName(name) => name.as_bytes_with_nul().len(),
            FamilyId(_id) => 2,
            Unknown(_, buffer) => buffer.len(),
        }
    }

    fn kind(&self) -> u16 {
        use ControlNla::*;
        match self {
            FamilyName(_) => libc::CTRL_ATTR_FAMILY_NAME as u16,
            FamilyId(_) => libc::CTRL_ATTR_FAMILY_ID as u16,
            Unknown(kind, _) => *kind,
        }
    }

    fn emit_value(&self, mut buffer: &mut [u8]) {
        use ControlNla::*;
        match self {
            FamilyName(name) => {
                let _ = buffer.write(name.as_bytes()).unwrap();
            }
            FamilyId(id) => {
                NativeEndian::write_u16(buffer, *id);
            }

            Unknown(_, value) => {
                let _ = buffer.write(value).unwrap();
            }
        }
    }
}

impl<'a, T: AsRef<[u8]> + 'a + ?Sized + std::fmt::Debug> Parseable<NlaBuffer<&'a T>>
    for ControlNla
{
    fn parse(buf: &NlaBuffer<&'a T>) -> Result<Self, DecodeError> {
        let nla = match buf.kind() as i32 {
            libc::CTRL_ATTR_FAMILY_NAME => {
                ControlNla::FamilyName(parsers::parse_cstring(buf.value())?)
            }
            libc::CTRL_ATTR_FAMILY_ID => ControlNla::FamilyId(parsers::parse_u16(buf.value())?),
            _unknown_kind => ControlNla::Unknown(buf.kind(), buf.value().to_vec()),
        };
        Ok(nla)
    }
}
