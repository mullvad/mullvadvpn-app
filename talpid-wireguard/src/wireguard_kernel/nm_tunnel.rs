use super::{
    super::stats::{Stats, StatsMap},
    Config, Error as WgKernelError, Handle, Tunnel, TunnelError, MULLVAD_INTERFACE_NAME,
};
use futures::Future;
use std::{collections::HashMap, pin::Pin};
use talpid_dbus::{
    dbus,
    network_manager::{
        DeviceConfig, Error as NetworkManagerError, NetworkManager, Variant, VariantMap,
        WireguardTunnel,
    },
};

#[derive(err_derive::Error, Debug)]
pub enum Error {
    #[error(display = "Error while communicating over Dbus")]
    Dbus(#[error(source)] dbus::Error),

    #[error(display = "NetworkManager error")]
    NetworkManager(#[error(source)] NetworkManagerError),
}

pub struct NetworkManagerTunnel {
    network_manager: NetworkManager,
    tunnel: Option<WireguardTunnel>,
    netlink_connections: Handle,
    tokio_handle: tokio::runtime::Handle,
    interface_name: String,
}

impl NetworkManagerTunnel {
    pub fn new(
        tokio_handle: tokio::runtime::Handle,
        config: &Config,
    ) -> std::result::Result<Self, WgKernelError> {
        let network_manager = NetworkManager::new()
            .map_err(Error::NetworkManager)
            .map_err(WgKernelError::NetworkManager)?;
        let config_map = convert_config_to_dbus(config);
        let tunnel = network_manager
            .create_wg_tunnel(&config_map)
            .map_err(|err| WgKernelError::NetworkManager(err.into()))?;

        let interface_name = match network_manager.get_interface_name(&tunnel) {
            Ok(name) => name,
            Err(error) => {
                log::error!("Failed to fetch interface name from NM: {}", error);
                MULLVAD_INTERFACE_NAME.to_string()
            }
        };
        let netlink_connections = tokio_handle.block_on(Handle::connect())?;

        Ok(NetworkManagerTunnel {
            network_manager,
            tunnel: Some(tunnel),
            netlink_connections,
            tokio_handle,
            interface_name,
        })
    }
}

impl Tunnel for NetworkManagerTunnel {
    fn get_interface_name(&self) -> String {
        self.interface_name.clone()
    }

    fn stop(mut self: Box<Self>) -> std::result::Result<(), TunnelError> {
        if let Some(tunnel) = self.tunnel.take() {
            if let Err(err) = self.network_manager.remove_tunnel(tunnel) {
                log::error!("Failed to remove WireGuard tunnel via NM: {}", err);
                Err(TunnelError::StopWireguardError { status: 0 })
            } else {
                Ok(())
            }
        } else {
            Ok(())
        }
    }

    fn get_tunnel_stats(&self) -> std::result::Result<StatsMap, TunnelError> {
        let mut wg = self.netlink_connections.wg_handle.clone();
        self.tokio_handle.block_on(async move {
            let device = wg
                .get_by_name(self.interface_name.clone())
                .await
                .map_err(|err| {
                    log::error!("Failed to fetch WireGuard device config: {}", err);
                    TunnelError::GetConfigError
                })?;
            Ok(Stats::parse_device_message(&device))
        })
    }

    fn set_config(
        &self,
        config: Config,
    ) -> Pin<Box<dyn Future<Output = std::result::Result<(), TunnelError>> + Send>> {
        let interface_name = self.interface_name.clone();
        let mut wg = self.netlink_connections.wg_handle.clone();
        Box::pin(async move {
            let index = iface_index(&interface_name).map_err(|err| {
                log::error!("Failed to fetch WireGuard device index: {}", err);
                TunnelError::SetConfigError
            })?;
            wg.set_config(index, &config).await.map_err(|err| {
                log::error!("Failed to apply WireGuard config: {}", err);
                TunnelError::SetConfigError
            })
        })
    }
}

fn convert_config_to_dbus(config: &Config) -> DeviceConfig {
    let mut ipv6_config: VariantMap = HashMap::new();
    let mut ipv4_config: VariantMap = HashMap::new();
    let mut wireguard_config: VariantMap = HashMap::new();
    let mut connection_config: VariantMap = HashMap::new();
    let mut peer_configs = vec![];

    wireguard_config.insert("mtu".into(), Variant(Box::new(config.mtu as u32)));
    if let Some(fwmark) = config.fwmark {
        wireguard_config.insert("fwmark".into(), Variant(Box::new(fwmark as u32)));
    }
    wireguard_config.insert("peer-routes".into(), Variant(Box::new(false)));
    wireguard_config.insert(
        "private-key".into(),
        Variant(Box::new(config.tunnel.private_key.to_base64())),
    );
    wireguard_config.insert("private-key-flags".into(), Variant(Box::new(0x0u32)));

    for peer in config.peers.iter() {
        let mut peer_config: VariantMap = HashMap::new();
        let allowed_ips = peer
            .allowed_ips
            .iter()
            .map(ToString::to_string)
            .collect::<Vec<_>>();

        peer_config.insert("allowed-ips".into(), Variant(Box::new(allowed_ips)));
        peer_config.insert(
            "endpoint".into(),
            Variant(Box::new(peer.endpoint.to_string())),
        );
        peer_config.insert(
            "public-key".into(),
            Variant(Box::new(peer.public_key.to_base64())),
        );

        peer_configs.push(peer_config);
    }
    wireguard_config.insert("peers".into(), Variant(Box::new(peer_configs)));

    connection_config.insert("type".into(), Variant(Box::new("wireguard".to_string())));
    connection_config.insert(
        "id".into(),
        Variant(Box::new(MULLVAD_INTERFACE_NAME.to_string())),
    );
    connection_config.insert(
        "interface-name".into(),
        Variant(Box::new(MULLVAD_INTERFACE_NAME.to_string())),
    );
    connection_config.insert("autoconnect".into(), Variant(Box::new(true)));

    let ipv4_addrs: Vec<_> = config
        .tunnel
        .addresses
        .iter()
        .filter(|ip| ip.is_ipv4())
        .map(NetworkManager::convert_address_to_dbus)
        .collect();

    let ipv6_addrs: Vec<_> = config
        .tunnel
        .addresses
        .iter()
        .filter(|ip| ip.is_ipv6())
        .map(NetworkManager::convert_address_to_dbus)
        .collect();

    ipv4_config.insert("address-data".into(), Variant(Box::new(ipv4_addrs)));
    ipv4_config.insert("ignore-auto-routes".into(), Variant(Box::new(true)));
    ipv4_config.insert("ignore-auto-dns".into(), Variant(Box::new(true)));
    ipv4_config.insert("may-fail".into(), Variant(Box::new(true)));
    ipv4_config.insert("method".into(), Variant(Box::new("manual".to_string())));
    ipv4_config.insert("never-default".into(), Variant(Box::new(true)));

    if !ipv6_addrs.is_empty() {
        ipv6_config.insert("method".into(), Variant(Box::new("manual".to_string())));
        ipv6_config.insert("address-data".into(), Variant(Box::new(ipv6_addrs)));
        ipv6_config.insert("ignore-auto-routes".into(), Variant(Box::new(true)));
        ipv6_config.insert("ignore-auto-dns".into(), Variant(Box::new(true)));
        ipv6_config.insert("may-fail".into(), Variant(Box::new(true)));
    }

    let mut settings = HashMap::new();
    settings.insert("ipv4".into(), ipv4_config);
    if !ipv6_config.is_empty() {
        settings.insert("ipv6".into(), ipv6_config);
    }
    settings.insert("wireguard".into(), wireguard_config);
    settings.insert("connection".into(), connection_config);

    settings
}

/// Converts an interface name into the corresponding index.
#[cfg(target_os = "linux")]
fn iface_index(name: &str) -> std::result::Result<libc::c_uint, IfaceIndexLookupError> {
    let c_name = std::ffi::CString::new(name)
        .map_err(|e| IfaceIndexLookupError::InvalidInterfaceName(name.to_owned(), e))?;
    let index = unsafe { libc::if_nametoindex(c_name.as_ptr()) };
    if index == 0 {
        Err(IfaceIndexLookupError::InterfaceLookupError(
            name.to_owned(),
            std::io::Error::last_os_error(),
        ))
    } else {
        Ok(index)
    }
}

/// Failure to lookup an interfaces index by its name.
#[derive(Debug, err_derive::Error)]
pub enum IfaceIndexLookupError {
    /// The interface name is invalid -  contains null bytes or is too long.
    #[error(display = "Invalid network interface name: {}", _0)]
    InvalidInterfaceName(String, #[error(source)] std::ffi::NulError),
    /// Interface wasn't found by its name.
    #[error(display = "Failed to get index for interface {}", _0)]
    InterfaceLookupError(String, #[error(source)] std::io::Error),
}
