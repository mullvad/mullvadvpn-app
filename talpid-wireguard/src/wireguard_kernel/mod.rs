use super::{Config, Tunnel, TunnelError};
use futures::future::{abortable, AbortHandle};
use netlink_packet_core::{constants::*, NetlinkDeserializable};
use netlink_packet_route::{
    rtnl::{
        address::nlas::Nla as AddressNla,
        link::nlas::{Info, InfoKind, Nla as LinkNla},
        AddressMessage, LinkMessage, RtnlMessage, RT_SCOPE_UNIVERSE,
    },
    NetlinkMessage, NetlinkPayload,
};
use netlink_packet_utils::DecodeError;
use netlink_proto::{
    sys::{protocols::NETLINK_GENERIC, SocketAddr},
    ConnectionHandle, Error as NetlinkError,
};
use std::{ffi::CString, net::IpAddr};
use tokio_stream::StreamExt;

mod parsers;

pub mod wg_message;
use wg_message::{DeviceMessage, DeviceNla};
pub mod nl_message;
use nl_message::{ControlNla, NetlinkControlMessage};

pub mod netlink_tunnel;
pub use netlink_tunnel::NetlinkTunnel;
pub mod nm_tunnel;
pub use nm_tunnel::NetworkManagerTunnel;

#[derive(err_derive::Error, Debug)]
#[error(no_from)]
pub enum Error {
    #[error(display = "Failed to decode netlink message")]
    Decode(#[error(source)] DecodeError),

    #[error(display = "Failed to execute netlink control request")]
    NetlinkControlMessage(#[error(source)] nl_message::Error),

    #[error(display = "Failed to open netlink socket")]
    NetlinkSocket(#[error(source)] std::io::Error),

    #[error(display = "Failed to send netlink control request")]
    NetlinkRequest(#[error(source)] netlink_proto::Error<NetlinkControlMessage>),

    #[error(display = "WireGuard netlink interface unavailable. Is the kernel module loaded?")]
    WireguardNetlinkInterfaceUnavailable,

    #[error(display = "Unknown WireGuard command _0")]
    UnnkownWireguardCommmand(u8),

    #[error(display = "Received no response")]
    NoResponse,

    #[error(display = "Received truncated message")]
    Truncated,

    #[error(display = "WireGuard device does not exist")]
    NoDevice,

    #[error(display = "Failed to get config: _0")]
    WgGetConf(netlink_packet_core::error::ErrorMessage),

    #[error(display = "Failed to apply config: _0")]
    WgSetConf(netlink_packet_core::error::ErrorMessage),

    #[error(display = "Interface name too long")]
    InterfaceName,

    #[error(display = "Send request error")]
    SendRequest(#[error(source)] NetlinkError<DeviceMessage>),

    #[error(display = "Create device error")]
    NetlinkCreateDevice(#[error(source)] rtnetlink::Error),

    #[error(display = "Add IP to device error")]
    NetlinkSetIp(rtnetlink::Error),

    #[error(display = "Failed to delete device")]
    DeleteDevice(#[error(source)] rtnetlink::Error),

    #[error(display = "NetworkManager error")]
    NetworkManager(#[error(source)] nm_tunnel::Error),
}

pub(crate) const MULLVAD_INTERFACE_NAME: &str = "wg-mullvad";

#[derive(Debug)]
pub struct Handle {
    pub wg_handle: WireguardConnection,
    route_handle: rtnetlink::Handle,
    wg_abort_handle: AbortHandle,
    route_abort_handle: AbortHandle,
}

impl Handle {
    pub async fn connect() -> Result<Self, Error> {
        let message_type = Self::get_wireguard_message_type().await?;
        let (conn, wireguard_connection, _messages) =
            netlink_proto::new_connection(NETLINK_GENERIC).map_err(Error::NetlinkSocket)?;
        let wg_handle = WireguardConnection {
            message_type,
            connection: wireguard_connection,
        };
        let (abortable_connection, wg_abort_handle) = abortable(conn);
        tokio::spawn(abortable_connection);
        let (conn, route_handle, _messages) =
            rtnetlink::new_connection().map_err(Error::NetlinkSocket)?;
        let (abortable_connection, route_abort_handle) = abortable(conn);
        tokio::spawn(abortable_connection);

        Ok(Self {
            wg_handle,
            route_handle,
            wg_abort_handle,
            route_abort_handle,
        })
    }

    async fn get_wireguard_message_type() -> Result<u16, Error> {
        let (conn, mut handle, _messages) =
            netlink_proto::new_connection(NETLINK_GENERIC).map_err(Error::NetlinkSocket)?;
        let (conn, abort_handle) = abortable(conn);
        tokio::spawn(conn);

        let result = async move {
            let mut message: NetlinkMessage<NetlinkControlMessage> =
                NetlinkControlMessage::get_netlink_family_id(CString::new("wireguard").unwrap())
                    .map_err(Error::NetlinkControlMessage)?
                    .into();

            message.header.flags = NLM_F_REQUEST | NLM_F_ACK;

            let mut req = handle
                .request(message, SocketAddr::new(0, 0))
                .map_err(Error::NetlinkRequest)?;
            let response = req.next().await;
            if let Some(response) = response {
                if let NetlinkPayload::InnerMessage(msg) = response.payload {
                    for nla in msg.nlas.into_iter() {
                        if let ControlNla::FamilyId(id) = nla {
                            return Ok(id);
                        }
                    }
                }
            }
            Err(Error::WireguardNetlinkInterfaceUnavailable)
        }
        .await;

        abort_handle.abort();
        result
    }

    // create a wireguard device with the given name.
    pub async fn create_device(&mut self, name: String, mtu: u32) -> Result<u32, Error> {
        let mut message = LinkMessage::default();

        // set link to be up
        message.header.flags = netlink_packet_route::IFF_UP;
        // message.header.change_mask = netlink_packet_route::IFF_UP;
        // set link name
        message.nlas.push(LinkNla::IfName(name.clone()));
        // set link MTU
        message.nlas.push(LinkNla::Mtu(mtu));
        // set link type
        message
            .nlas
            .push(LinkNla::Info(vec![Info::Kind(InfoKind::Other(
                "wireguard".to_string(),
            ))]));

        let mut add_request = NetlinkMessage::from(RtnlMessage::NewLink(message));
        add_request.header.flags =
            NLM_F_REQUEST | NLM_F_ACK | NLM_F_REPLACE | NLM_F_CREATE | NLM_F_MATCH;
        let mut response = self
            .route_handle
            .request(add_request)
            .map_err(Error::NetlinkCreateDevice)?;
        while let Some(response_message) = response.next().await {
            if let NetlinkPayload::Error(err) = response_message.payload {
                // if the device exists, verify that it's a wireguard device
                if -err.code != libc::EEXIST {
                    return Err(Error::NetlinkCreateDevice(rtnetlink::Error::NetlinkError(
                        err,
                    )));
                }
            }
        }

        // fetch interface index of new device
        let new_device = self.wg_handle.get_by_name(name).await?;
        for nla in new_device.nlas {
            if let DeviceNla::IfIndex(index) = nla {
                return Ok(index);
            }
        }

        Err(Error::NoDevice)
    }

    pub async fn set_ip_address(&mut self, index: u32, addr: IpAddr) -> Result<(), Error> {
        let address_message = add_ip_addr_message(index, addr);
        let mut request = NetlinkMessage::from(RtnlMessage::NewAddress(address_message));
        request.header.flags = NLM_F_REQUEST | NLM_F_ACK | NLM_F_CREATE | NLM_F_REPLACE;

        let mut response = self
            .route_handle
            .request(request)
            .map_err(Error::NetlinkSetIp)?;
        while let Some(response_message) = response.next().await {
            consume_netlink_error(response_message, Error::NetlinkSetIp)?;
        }

        Ok(())
    }

    pub async fn delete_device(&mut self, index: u32) -> Result<(), Error> {
        let mut link_message = LinkMessage::default();
        link_message.header.index = index;

        let mut request = NetlinkMessage::from(RtnlMessage::DelLink(link_message));
        request.header.flags = NLM_F_REQUEST | NLM_F_ACK;

        let mut response = self
            .route_handle
            .request(request)
            .map_err(Error::DeleteDevice)?;
        while let Some(message) = response.next().await {
            consume_netlink_error(message, Error::DeleteDevice)?;
        }

        Ok(())
    }
}

impl Drop for Handle {
    fn drop(&mut self) {
        self.wg_abort_handle.abort();
        self.route_abort_handle.abort();
    }
}

#[derive(Debug, Clone)]
pub struct WireguardConnection {
    connection: ConnectionHandle<DeviceMessage>,
    message_type: u16,
}

impl WireguardConnection {
    pub async fn get_by_name(&mut self, name: String) -> Result<DeviceMessage, Error> {
        self.fetch_device(DeviceMessage::get_by_name(self.message_type, name)?)
            .await
    }

    pub async fn get_by_index(&mut self, index: u32) -> Result<DeviceMessage, Error> {
        self.fetch_device(DeviceMessage::get_by_index(self.message_type, index))
            .await
    }

    pub async fn fetch_device(
        &mut self,
        device_message: DeviceMessage,
    ) -> Result<DeviceMessage, Error> {
        let mut netlink_message = NetlinkMessage::from(device_message);
        netlink_message.header.flags = NLM_F_REQUEST | NLM_F_ACK | NLM_F_DUMP;

        let mut response = self
            .connection
            .request(netlink_message, SocketAddr::new(0, 0))
            .map_err(Error::SendRequest)?;
        match response.next().await {
            Some(received_message) => match received_message.payload {
                NetlinkPayload::InnerMessage(inner) => Ok(inner),
                NetlinkPayload::Error(err) => {
                    if err.code == -libc::ENODEV {
                        Err(Error::NoDevice)
                    } else {
                        Err(Error::WgGetConf(err))
                    }
                }
                anything_else => {
                    log::error!("Received unexpected response: {:?}", anything_else);
                    Err(Error::NoResponse)
                }
            },
            None => Err(Error::NoResponse),
        }
    }

    pub async fn set_config(&mut self, interface_index: u32, config: &Config) -> Result<(), Error> {
        let message = DeviceMessage::reset_config(self.message_type, interface_index, config);
        let mut netlink_message = NetlinkMessage::from(message);
        netlink_message.header.flags = NLM_F_REQUEST | NLM_F_ACK;

        let mut request = self
            .connection
            .request(netlink_message, SocketAddr::new(0, 0))
            .map_err(Error::SendRequest)?;

        while let Some(response) = request.next().await {
            if let NetlinkPayload::Error(err) = response.payload {
                return Err(Error::WgSetConf(err));
            }
        }
        Ok(())
    }
}

fn consume_netlink_error<
    I: NetlinkDeserializable + Clone + Eq + std::fmt::Debug,
    F: Fn(rtnetlink::Error) -> Error,
>(
    message: NetlinkMessage<I>,
    err_constructor: F,
) -> Result<(), Error> {
    if let NetlinkPayload::Error(err) = message.payload {
        return Err(err_constructor(rtnetlink::Error::NetlinkError(err)));
    }
    Ok(())
}

// the built-in support for adding addresses is too helpful, so a simple AddressMessage with a
// single Address nla is created
fn add_ip_addr_message(if_index: u32, addr: IpAddr) -> AddressMessage {
    let prefix_len = if addr.is_ipv4() { 32 } else { 128 };
    let mut message = AddressMessage::default();
    message.header.prefix_len = prefix_len;
    message.header.index = if_index;
    message.header.scope = RT_SCOPE_UNIVERSE;

    match addr {
        IpAddr::V4(ipv4) => {
            message.header.family = libc::AF_INET as u8;
            let ip_bytes = ipv4.octets().to_vec();

            message.nlas.push(AddressNla::Address(ip_bytes.clone()));
            message.nlas.push(AddressNla::Local(ip_bytes));
        }
        IpAddr::V6(ipv6) => {
            message.header.family = libc::AF_INET6 as u8;
            message
                .nlas
                .push(AddressNla::Address(ipv6.octets().to_vec()));
        }
    };

    message
}
