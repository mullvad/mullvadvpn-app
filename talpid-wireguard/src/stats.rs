#[cfg(target_os = "linux")]
use super::wireguard_kernel::wg_message::{DeviceMessage, DeviceNla, PeerNla};

#[derive(err_derive::Error, Debug, PartialEq)]
pub enum Error {
    #[error(display = "Failed to parse peer pubkey from string \"_0\"")]
    PubKeyParse(String, #[error(source)] hex::FromHexError),

    #[error(display = "Failed to parse integer from string \"_0\"")]
    IntParse(String, #[error(source)] std::num::ParseIntError),

    #[error(display = "Device no longer exists")]
    NoTunnelDevice,

    #[error(display = "Failed to obtain tunnel config")]
    NoTunnelConfig,
}

/// Contains bytes sent and received through a tunnel
#[derive(Default, Debug, PartialEq, Eq, Clone, Copy)]
pub struct Stats {
    pub tx_bytes: u64,
    pub rx_bytes: u64,
}

/// A map from peer pubkeys to peer stats.
pub type StatsMap = std::collections::HashMap<[u8; 32], Stats>;

impl Stats {
    pub fn parse_config_str(config: &str) -> Result<StatsMap, Error> {
        let mut map = StatsMap::new();

        let mut peer = None;
        let mut tx_bytes = None;
        let mut rx_bytes = None;

        // parts iterates over keys and values
        let parts = config.split('\n').filter_map(|line| {
            let mut pair = line.split('=');
            let key = pair.next()?;
            let value = pair.next()?;
            Some((key, value))
        });

        for (key, value) in parts {
            match key {
                "public_key" => {
                    let mut buffer = [0u8; 32];
                    hex::decode_to_slice(value, &mut buffer)
                        .map_err(|err| Error::PubKeyParse(value.to_string(), err))?;
                    peer = Some(buffer);
                    tx_bytes = None;
                    rx_bytes = None;
                }
                "rx_bytes" => {
                    rx_bytes = Some(
                        value
                            .trim()
                            .parse()
                            .map_err(|err| Error::IntParse(value.to_string(), err))?,
                    );
                }
                "tx_bytes" => {
                    tx_bytes = Some(
                        value
                            .trim()
                            .parse()
                            .map_err(|err| Error::IntParse(value.to_string(), err))?,
                    );
                }

                _ => continue,
            }

            if let (Some(peer_val), Some(tx_bytes_val), Some(rx_bytes_val)) =
                (peer, tx_bytes, rx_bytes)
            {
                map.insert(
                    peer_val,
                    Self {
                        tx_bytes: tx_bytes_val,
                        rx_bytes: rx_bytes_val,
                    },
                );
                peer = None;
                tx_bytes = None;
                rx_bytes = None;
            }
        }
        Ok(map)
    }

    #[cfg(target_os = "linux")]
    pub fn parse_device_message(message: &DeviceMessage) -> StatsMap {
        let mut map = StatsMap::new();

        for nla in &message.nlas {
            if let DeviceNla::Peers(peers) = nla {
                for msg in peers {
                    let mut tx_bytes = 0;
                    let mut rx_bytes = 0;
                    let mut pub_key = None;

                    for nla in &msg.0 {
                        match nla {
                            PeerNla::TxBytes(bytes) => tx_bytes = *bytes,
                            PeerNla::RxBytes(bytes) => rx_bytes = *bytes,
                            PeerNla::PublicKey(key) => pub_key = Some(*key),
                            _ => continue,
                        }
                    }
                    if let Some(key) = pub_key {
                        map.insert(key, Stats { tx_bytes, rx_bytes });
                    }
                }
            }
        }

        map
    }
}

#[cfg(test)]
mod test {
    use super::{Error, Stats};

    #[test]
    fn test_parsing() {
        let valid_input = "private_key=0000000000000000000000000000000000000000000000000000000000000000\npublic_key=0000000000000000000000000000000000000000000000000000000000000000\npreshared_key=0000000000000000000000000000000000000000000000000000000000000000\nprotocol_version=1\nendpoint=000.000.000.000:00000\nlast_handshake_time_sec=1578420649\nlast_handshake_time_nsec=369416131\ntx_bytes=2740\nrx_bytes=2396\npersistent_keepalive_interval=0\nallowed_ip=0.0.0.0/0\n";
        let pubkey = [0u8; 32];

        let stats = Stats::parse_config_str(valid_input).expect("Failed to parse valid input");
        assert_eq!(stats.len(), 1);
        let actual_keys: Vec<[u8; 32]> = stats.keys().cloned().collect();
        assert_eq!(actual_keys, [pubkey]);
        assert_eq!(stats[&pubkey].rx_bytes, 2396);
        assert_eq!(stats[&pubkey].tx_bytes, 2740);
    }

    #[test]
    fn test_parsing_invalid_input() {
        let invalid_input = "private_key=0000000000000000000000000000000000000000000000000000000000000000\npublic_key=0000000000000000000000000000000000000000000000000000000000000000\npreshared_key=0000000000000000000000000000000000000000000000000000000000000000\nprotocol_version=1\nendpoint=000.000.000.000:00000\nlast_handshake_time_sec=1578420649\nlast_handshake_time_nsec=369416131\ntx_bytes=27error40\npersistent_keepalive_interval=0\nallowed_ip=0.0.0.0/0\n";
        let invalid_str = "27error40".to_string();
        let int_err = invalid_str.parse::<u64>().unwrap_err();

        assert_eq!(
            Stats::parse_config_str(invalid_input),
            Err(Error::IntParse(invalid_str, int_err))
        );
    }
}
