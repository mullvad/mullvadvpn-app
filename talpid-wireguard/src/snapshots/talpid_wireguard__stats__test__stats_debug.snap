---
source: talpid-wireguard/src/stats.rs
expression: "StatsDebug { now, stats: &stats }"
---
Stats {
    tx_bytes: 100,
    rx_bytes: 100,
    last_handshake: "60000 ms ago",
    daita: None,
}
